product-classification/product_classification/InferenceEngine.py

```diff
@@ -0,0 +1,254 @@
+
+        if self.model is not None:
+            probas = self.model.predict_proba(df)
+            return probas
+        else:
+            # Handle the case when the model fails to load
+            logger.warning("Model failed to load. Returning empty DataFrame.")
+            return pd.DataFrame()
+
+    @staticmethod
+    def join_features_with_inference(
+        product_df: pd.DataFrame, classification_df: pd.DataFrame
+    ) -> pd.DataFrame:
+        """
+        This function is used to join the product dataframe with the retrieved classification
+        Dataframe.
+
+        Args:
+            product_df (pd.Dataframe): Dataframe with product instances.
+            classification_df (pd.Dataframe): Dataframe with ProductClassification instances.
+
+        Returns:
+            pd.DataFrame: pd.DataFrame of ProductClassification objects with thresholded
+            categories and corresponding thresholds.
+        """
+
+        return classification_df.join(product_df, on="id")
+
+    def apply_shop_thresholds(
+        self,
+        classification_df: pd.DataFrame,
+        product_df: Optional[pd.DataFrame] = None,
+    ) -> pd.DataFrame:
+        """
+        This function is used to apply thresholds on the predictions to exclude out-of-distribution
+        products.
+
+        Args:
+            product_df (pd.Dataframe): Dataframe with product instances.
+            classification_df (pd.Dataframe): Dataframe with ProductClassification instances.
+
+        Returns:
+            pd.DataFrame: pd.DataFrame of ProductClassification objects with thresholded
+            categories and corresponding thresholds.
+        """
+
+        # add fallback threshold if a shop/category was not present during evaluation
+        # use smallest threshold from all shops as fallback
+        fallback_thresh_by_category = (
+            self.shop_thresholds.groupby("predicted_category")["threshold"].agg(min).to_dict()
+        )
+        fallback_thresholds = classification_df["predicted_category"].apply(
+            lambda x: fallback_thresh_by_category.get(x)
+        )
+
+        # set shop specific thresholds if source and merchant are sent along with request
+        if product_df is not None and {"source", "merchant"}.issubset(set(product_df.columns)):
+            combined = self.join_features_with_inference(product_df, classification_df)
+            join_keys = ["ml_model_name", "source", "merchant", "predicted_category"]
+            combined = combined.join(self.shop_thresholds.set_index(join_keys), on=join_keys)
+            combined["threshold"].fillna(fallback_thresholds)
+        else:
+            combined = pd.concat(
+                [classification_df, fallback_thresholds.rename("threshold")], axis=1
+            )
+
+        combined["category_thresholded"] = combined.apply(
+            lambda x: x["predicted_category"]
+            if x["confidence"] >= x["threshold"]
+            else "under_threshold",
```

Comment:
Maybe you could explain the logic behind the under_threshold and what happens to those products